### DISCUSSION TITLE
ESP32-S3 based Adafruit QT Py S3 with 4MB FLASH and 2MB PSRAM not showing file system

### DISCUSSION THREAD

**Discussion by liudr** [micropython/micropython#16246] [ESP32] [answered]
Here is the board I have:

https://www.adafruit.com/product/5700

It has 4MB FLASH and 2MB PSRAM. I followed the instructions below to flash the generic firmware that claims it supports 4/8MB flash and optional PSRAM (V 1.23):

https://micropython.org/download/ESP32_GENERIC_S3/

After flashing, I was able to run MicroPython and imported gc and os to test free RAM and file space. PSRAM showed up fine but file system is zero bytes. I also tried to use Thonny to display file storage and also got all zeroes. 
![image](https://github.com/user-attachments/assets/0d655a9e-7944-44cb-88b6-7cd6b73c6480)

statvfs('/') returns (0, 0, 0, 0, 0, 0, 0, 0, 0, 128)

Trying to upload files to the board triggers the flash not initialized exception:
![image](https://github.com/user-attachments/assets/0b9e6177-8db3-4447-a8b7-7b5bb65f5d66)

So I wonder what went wrong. I also installed MP using Thonny. Log here showing commands:

Downloading from https://micropython.org/resources/firmware/ESP32_GENERIC_S3-20240602-v1.23.0.bin
Writing to C:\Users\user\AppData\Local\Temp\tmp0_qxtfkq\ESP32_GENERIC_S3-20240602-v1.23.0.bin
100%Disconnecting from REPL...C:\Users\user\AppData\Local\Programs\Thonny\python.exe -u -m esptool --port COM8 --chip esp32s3 --baud 115200 write_flash --flash_mode keep --flash_size keep --erase-all 0x0 C:\Users\user\AppData\Local\Temp\tmp0_qxtfkq\ESP32_GENERIC_S3-20240602-v1.23.0.bin
esptool.py v4.7.0
Serial port COM8
Connecting...
Chip is ESP32-S3 (QFN56) (revision v0.1)
Features: WiFi, BLE, Embedded Flash 4MB (XMC), Embedded PSRAM 2MB (AP_3v3)
Crystal is 40MHz
MAC: f4:12:fa:5a:2e:fc
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Erasing flash (this may take a while)...
Chip erase completed successfully in 17.0s
Compressed 1628144 bytes to 1067824...
Writing at 0x00000000... (1 %)
Writing at 0x00186666... (98 %)
Writing at 0x0018c39f... (100 %)
Wrote 1628144 bytes (1067824 compressed) at 0x00000000 in 21.3 seconds (effective 610.9 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
Done!

I'm sure this must be a beginner's mistake somehow. I've used ESP32 extensively but this is the first time I tried esp32-s3. I don't like the fact that resetting the board resets the serial port and causes thonny to lose connection so I've never bothered getting s3 boards until now.

**Comment by jonnor** [comment]
Try to erase the chip before flashing the firmware.

**Reply by liudr** [reply]
Did that several times and thonny also erases (see log) and didn't work.

**Comment by scruss** [comment, marked as answer]
Maybe try the [Firmware (4MiB flash)](https://micropython.org/download/ESP32_GENERIC_S3/) version (scroll down a bit) instead of the generic S3 one. 

You may still have the Adafruit uf2 bootloader on the board: try double-tapping the reset button. I can't remember if this particular board needs a short pause between taps or not

**Reply by liudr** [reply]
Double clicking has no effect (not that I know what to expect). According to adafruit, it may enumerate a drive but didn't do anything other than resetting on my board. But flashing with your recommended .bin image worked. So what is the difference between "Firmware" and "Firmware (4MiB flash)"? Thanks.
P.S. I'm doing this for teaching and teaching teachers. Running MP on S2/S3 may be more trouble than worth it since ESP32 boards are plenty.

**Reply by scruss** [reply]
> So what is the difference between "Firmware" and "Firmware (4MiB flash)"? 

There are several versions of the S3 chip. The flash/PSRAM detection isn't perfect, and there needs to be different configurations to pick up internal/external flash and RAM.

> Running MP on S2/S3 may be more trouble than worth it since ESP32 boards are plenty.

I feel your pain, but ESP32 (neither S2 nor S3) have even more confusing variants than the S-series chips. Maybe best *not* to flash from Thonny, as it may make assumptions about the hardware that aren't quite correct

**Reply by liudr** [reply]
I used both Thonny and esptool but since I used the wrong firmware for the board neither worked.

For ESP32, the worst you could do is to only use 4MB of FLASH with the SPIRAM variant of firmware image. I've not compiled the firmware since version 1.20 so I don't have the toolchain installed. It's a simple task to just clone a board definition and make a new partition table for 8MB, as long as you take the pain to install the espressif toolchain and clone MP repo. So I just go with 4MB FLASH firmware even though I may have an adafruit board with 8MB FLASH. I wish MP compiles a generic version of 8MB FLASH with SPIRAM. Adafruit produces a handful of boards with ESP32 that comes with 8MB FLASH (Feather ESP32, Feather ESP32 V2, QT Py ESP32 Pico etc. and Espressif's own WROVER dev kit) and micropython.org doesn't have a single ESP32 variant compiled firmware for them. I've been using generic 4MB FLASH with SPIRAM for those boards.

ESP32 with 8MB FLASH:
- Adafruit ItsyBitsy ESP32
- Adafruit Feather ESP32
- Adafruit Feather ESP32 V2
- Adafruit QT Py ESP32 Pico
- Espressif WROVER dev kit
- Probably more boards.

**Reply by scruss** [reply]
let me introduce you to [mp-image-tool-esp32](https://github.com/glenn20/mp-image-tool-esp32), which will resize the fs for you. My [SparkFun Thing Plus SkeleBoard](https://www.sparkfun.com/products/retired/18581)s (*doot!*) have 16 MB of flash, and this is useful.

**Reply by liudr** [reply]
> let me introduce you to [mp-image-tool-esp32](https://github.com/glenn20/mp-image-tool-esp32), which will resize the fs for you. My [SparkFun Thing Plus SkeleBoard](https://www.sparkfun.com/products/retired/18581)s (_doot!_) have 16 MB of flash, and this is useful.

Excellent! Thanks for the links! I was able to resize an espressif WROVER dev kit to 6MB VFS. I had to first do a resize and then do a grow per the tool's suggestion to fix file system corruption. I guess I need to understand more about VFS to know why the second step was needed.
